pkg/gameservers/controller.go

```diff
@@ -543,23 +543,29 @@ func (c *Controller) syncDevelopmentGameServer(ctx context.Context, gs *agonesv1
 		return gs, nil
 	}
 
-	// Only move from Creating -> Ready. Other manual state changes are up to the end user.
-	// We also don't want to move from Allocated -> Ready every time someone allocates a GameServer.
-	if gs.Status.State != agonesv1.GameServerStateCreating {
+	// Only move from Creating -> Ready or RequestReady -> Ready.
```

Comment:
Can you explain the flow here?

Messing with GameServer state worries me a little 😄 , and I'm not entirely sure how this ends up redirecting traffic to the local gameserver + sdk instance? (or maybe it doesn't?)

I'm wondering if there's a way we could hijack the existing `agones.dev/dev-address` annotation to keep this as a special case flow? 🤔 

https://agones.dev/site/docs/guides/local-game-server/

Reply:
> Can you explain the flow here?

Note that this function only applies for dev game servers. Note this snippet called early in `syncDevelopmentGameServer` (this function):
```golang
if !isDevGs {
    return gs, nil
}
```
So this function ONLY applies to dev game servers.

Previously, this function ONLY progressed the dev game server from `Creating` to `Ready`. This change adds in the progression of `RequestReady` to `Ready`.

Why this is important: When working with the SDK, a call to `agonesSdk.ReadyAsync()` will put the `GameServer` into the `RequestReady` state. The `GameServer` must be in the `Ready` state for the allocation service (or alternative) to allocate the `GameServer` when requested. Progression from `RequestReady` to `Ready` is always done by the controller (this file), except for dev game servers (see `syncGameServerRequestReadyState()`). This change simply makes it always the case that `RequestReady` to `Ready` will properly be progressed.

Yes, it is true that a manual REST call could put the `GameServer` into `Ready`. Something like:
```
curl -X PUT -H "Content-Type: application/json" --data "$(curl -s http://localhost:8082/apis/agones.dev/v1/namespaces/default/gameservers/my-local-server | jq '.status.state="Ready"')" http://localhost:8082/apis/agones.dev/v1/namespaces/default/gameservers/my-local-server
```
But doing this as a separate step when developing is **_EXTREMELEY TEDIOUS_**. This also isn't what happens normally. If the normal (prod) flow takes care of this, then the controller should still handle it for dev servers as well.

> I'm wondering if there's a way we could hijack the existing `agones.dev/dev-address` annotation to keep this as a special case flow?

I do actually have thoughts about using annotations, similar to `agones.dev/dev-address`, for handling how state progression should work. I actually think that special cases should be minimized. IMHO, dev workflows should accomplish 2 things: be easy to do and representative of prod. Making controller-originating `GameServer` progression as similar between dev workflows and prod is important to make sure that realistic behavior is encountered during development/testing.

This all said, this change is very small -- simply adding in proper progression from `RequestReady` to `Ready` for dev servers. The changes that I have in mind that would try to bring more consistency and making use of annotations; that I'm saving for another PR (to keep this one more straight forward).

Reply:
aaaah I see I see! I didn't expand up and see that this was specific to dev servers 👍🏻 That makes total sense. Thanks for the context 👍🏻 

We should add some unit and/or e2e tests for this then as well.

https://github.com/googleforgames/agones/blob/c431327653809997d193722788b66b1c0b55354d/pkg/gameservers/controller_test.go#L171

https://github.com/googleforgames/agones/blob/c431327653809997d193722788b66b1c0b55354d/test/e2e/gameserver_test.go#L475

Reply:
Added the related tests in the suggested locations.